src/ccl_power.c

```diff
@@ -406,194 +443,123 @@ static void ccl_cosmology_compute_power_class(ccl_cosmology * cosmo, int * statu
-  int newstatus=0;
-  int pwstatus=0;
   
-  //If not, proceed
-  if(!*status){
-    
-    // After this loop x will contain log(k)
+  if (*status == 0) {
+    z=ccl_linlog_spacing(amin, ccl_splines->A_SPLINE_MIN_PK,
+			 amax, ccl_splines->A_SPLINE_NLOG_PK,
+			 ccl_splines->A_SPLINE_NA_PK);
+    if(z==NULL) {
+      *status = CCL_ERROR_MEMORY;
+      ccl_cosmology_set_status_message(cosmo,"ccl_power.c: ccl_cosmology_compute_power_class(): memory allocation\n");
+    }
+  }
+  if(*status==0) {
+    y2d_lin = malloc(nk * na * sizeof(double));
+    if(y2d_lin==NULL) {
+      *status = CCL_ERROR_MEMORY;
+      ccl_cosmology_set_status_message(cosmo,"ccl_power.c: ccl_cosmology_compute_power_class(): memory allocation\n");
+    }
+  }
+  if(*status==0) {
+    y2d_nl = malloc(nk * na * sizeof(double));
+    if(y2d_nl==NULL) {
+      *status = CCL_ERROR_MEMORY;
+      ccl_cosmology_set_status_message(cosmo,"ccl_power.c: ccl_cosmology_compute_power_class(): memory allocation\n");
+    }
+  }
+
+  if(*status==0) {
+    // After this loop x will contain log(k), y will contain log(P_nl), z will contain log(P_lin)
     // all in Mpc, not Mpc/h units!
-    double psout_l,ic;
+    double psout_l;
+    s=0;
     for (int i=0; i<nk; i++) {
       for (int j = 0; j < na; j++) {
 	//The 2D interpolation routines access the function values y_{k_ia_j} with the following ordering:
 	//y_ij = y2d[j*N_k + i]
 	//with i = 0,...,N_k-1 and j = 0,...,N_a-1.
-	newstatus |= spectra_pk_at_k_and_z(&ba, &pm, &sp,x[i],1./a[j]-1.+1e-10, &psout_l,&ic);
+	s |= spectra_pk_at_k_and_z(&ba, &pm, &sp,x[i],1./z[j]-1.+1e-10, &psout_l,&ic);
 	y2d_lin[j*nk+i] = log(psout_l);
       }
       x[i] = log(x[i]);
     }
-
-    
-    //If error, store status, we will free later
-    if(newstatus) {
+    if(s) {
       *status = CCL_ERROR_CLASS;
       ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): Error computing CLASS power spectrum\n");
     }
-
-      
   }
 
-  
-  //If no error, proceed
-  if(!*status) {
-    
-    gsl_spline2d * log_power = gsl_spline2d_alloc(PLIN_SPLINE_TYPE, nk,na);
-    pwstatus = gsl_spline2d_init(log_power, x, a, y2d_lin,nk,na);
-    
-    //If not, proceed
-    if(!pwstatus){
-      cosmo->data.p_lin = log_power;
-    } else {
-      gsl_spline2d_free(log_power);
-      *status = CCL_ERROR_SPLINE;
-      ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): Error creating log_power spline\n");
-    }
+  if(*status==0)
+    cosmo->data.p_lin=ccl_p2d_t_new(na,z,nk,x,y2d_lin,1,2,ccl_p2d_cclgrowth,1,NULL,0,ccl_p2d_3,status);
 
+  if(*status==0) {
+    // At the moment KMIN can't be less than CLASS's kmin in the nonlinear case.
+    if (kmin<(exp(sp.ln_k[0]))) {
+      *status = CCL_ERROR_CLASS;
+      ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): K_MIN is less than CLASS's kmin. Not yet supported for nonlinear P(k).\n");
+    }
   }
```

Comment:
Why are testing this if we set it above?

Reply:
No idea. This is a leftover from the previous code. Removed in my last commit